File: leetcode/347.top-k-frequent-elements/ts/topk.test.ts
import { test, expect } from "vitest"
import topKFrequent from "./topk.js"

interface testcase {
    nums: number[]
    k: number
    want: number[]
}

test("topk", () => {
    const tcs: testcase[] = [
        { nums: [1, 1, 1, 2, 2, 3], k: 2, want: [1, 2] },
        { nums: [1], k: 1, want: [] },
        { nums: [1, 2, 1, 2, 1, 2, 3, 1, 3, 2], k: 2, want: [1, 2] },
    ]

    tcs.forEach(tc => expect(topKFrequent(tc.nums, tc.k) === tc.want))
})

File: leetcode/347.top-k-frequent-elements/cpp/topk_test.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

#include "topk.cpp"

struct testCase
{
    std::vector<int> nums;
    int k;
    std::vector<int> want;
};

TEST_CASE("TopKFrequency")
{
    std::vector<testCase> tcs{
        {{1, 1, 1, 2, 2, 3}, 2, {1, 2}},
        {{1}, 1, {1}},
        {{1, 2, 1, 2, 1, 2, 3, 1, 3, 2}, 2, {2, 1}},
    };

    Solution s;

    for (auto tc : tcs)
    {
        auto got = s.topKFrequent(tc.nums, tc.k);
        CHECK(got == tc.want);
    }
}
